gui: Disable Restore Versions filters when no versioned files exist (fixes #5408)

tomasz1986 opened the pull request:
Currently, the name and date filters in the Restore Versions modal are always enabled, even if there are no versioned files present. With this change, they are enabled only when there are no errors and versioned files actually exist.

In addition to disabling the filters, also completely skip date picker generation, as it serves no function while still displaying a non-sense date range, which starts today and ends in the past.

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

### Screenshots

#### Before

![image](https://user-images.githubusercontent.com/5626656/190013612-b1d67cfc-3eac-461e-8133-f315d60379ed.png)

#### After

![image](https://user-images.githubusercontent.com/5626656/190013626-8de9ef53-7239-47cc-9810-0a30cd8ba09c.png)

calmh commented:
Is there really no sane way of checking if an object is empty? Do we need to create a helper function?

tomasz1986 commented:
Yeah, I actually marked this WIP, as I'm going to change the approach completely. There's no need to generate the tree at all when there're no versions.

tomasz1986 commented:
I've changed the code a bit, so that tree generation is skipped almost immediately if there are no versions. I don't think it's that great though, as ideally all further parsing should be stopped in the function below. There's no point going forward if `data` is just `{}`, but I'm not really sure how to change the code to do do it cleanly.

https://github.com/syncthing/syncthing/blob/39d3424e34ae9c4636c4da53a075f744c6776567/gui/default/syncthing/core/syncthingController.js#L2592-L2605

I've also modified the HTML, so that the table and filter inputs are always shown. This way there's no GUI jumping, as the modal stays the same size and doesn't shrink, and also the `Loading data...` line is shown inside the table and not before it. I've also changed `loading: $translate.instant("Loading...")` to `loading: $translate.instant("Loading data...")` in JS to make both the same for consistency.

Do you mean creating a helper function instead of `(restoreVersions.versions | json) === '{}'`?

calmh commented:
Yeah that's pretty ugly I think. There's `angular.equals(foo, {})` or `query.isEmptyObject` we could maybe use instead.

imsodin commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -2591,17 +2591,21 @@ angular.module('syncthing.core')
-                                    return b.versionTime - a.versionTime;
+                            if (isEmptyObject(data)) {
+                                $scope.restoreVersions.noversions = true;
+                            } else {
```
Could you initially set it to `$scope.restoreVersions.versions = null` and then `if !closed && !isEmptyObject { iterate and assign to ...versions`? Then in all the conditions you can check for `!restoreVersions.versions`. No new fields or weird stringify checks to identify an empty object.

tomasz1986 replied:
I've done that now (and actually did try doing something similar a few times before), but there's one problem that I'm unable to solve. Basically, with the current code, the Restore Versions modal loads fine on the first try. However, if you close and re-open it, the following error pops up.

```
TypeError: Cannot read properties of undefined (reading 'expanderClosed')
    at setIcon (jquery.fancytree-all-deps.js:11752:14)
    at Fancytree.nodeRenderStatus (jquery.fancytree-all-deps.js:11848:5)
    at Fancytree.nodeRenderStatus (jquery.fancytree-all-deps.js:1701:22)
    at Fancytree.nodeRenderTitle (jquery.fancytree-all-deps.js:6521:10)
    at Fancytree._super (jquery.fancytree-all-deps.js:1685:22)
    at Fancytree.nodeRenderTitle (jquery.fancytree-all-deps.js:13185:15)
    at Fancytree.nodeRenderTitle (jquery.fancytree-all-deps.js:1701:22)
    at Fancytree.nodeRender (jquery.fancytree-all-deps.js:13123:11)
    at Fancytree.nodeRender (jquery.fancytree-all-deps.js:1701:22)
    at Fancytree.nodeRender (jquery.fancytree-all-deps.js:13144:11)
```

I've only managed to get around it by changing the code from

```
        <table id="restoreTree">
          <tbody>
            <tr>
              <td>
                <span ng-if="restoreVersions.versions"translate>Loading data...</span>
                <span ng-if="!restoreVersions.versions" translate>There are no file versions to restore.</span>
              </td>
              <td></td>
            </tr>
          </tbody>
        </table>
```

to something more like

```
        <table id="restoreTree" ng-if="restoreVersions.versions">
          <tbody>
            <tr>
              <td translate>Loading data...</td>
              <td></td>
            </tr>
          </tbody>
        </table>
        <table id="restoreTree" ng-if="!restoreVersions.versions">
          <tbody>
            <tr>
              <td translate>There are no file versions to restore.</td>
              <td></td>
            </tr>
          </tbody>
        </table>
```

which does get rid of the error, but the table code ends up being duplicated. The two messages themselves need to be inside the table cells in order to match the same style (padding, etc.) as when there are actual versioned files rendered there.

imsodin replied:
Can you try readding the condition in L3:
https://github.com/syncthing/syncthing/pull/8539/commits/39249a02cac5c0fabb4b14190007f5bd7a9e56f5#diff-35607bd09efb510801fed738df57073a32758c8f16b7f40693fe007d92fd306dL3  
Not that I'd understand why it helps, but it looks like an unrelated change and with this UI code any change in code may trigger any change in behaviour (maybe not quite as bad, but almost :P ).

tomasz1986 replied:
Yeah, it works with that too (as it's essentially the same as when moved down to the table element), but the problem is that I wanted to move the span element

https://github.com/syncthing/syncthing/pull/8539/files#diff-35607bd09efb510801fed738df57073a32758c8f16b7f40693fe007d92fd306dL3

from outside into the table, so that its position and padding matches the table cells, which prevents the text from jumping that takes place now while the table is being loaded 😕. That's why I was experimenting with two different conditions — one for `restoreVersions.versions` itself, and one for when `restoreVersions.versions` is empty.

tomasz1986 commented:
I've found the helper function `sizeOf` which is already used in other HTML files of the GUI, so I've decided to use it here too, which has simplified the whole diff greatly. Please check the current iteration and let me know whether the code is fine like this or whether there's still something to fix.

imsodin commented on gui/default/syncthing/folder/restoreVersionsModalView.html:
```diff
@@ -1,12 +1,14 @@
           <tbody>
             <tr>
-              <td></td>
+              <td>
+                <span ng-if="sizeOf(restoreVersions.versions) > 0" translate>Loading data...</span>
+                <span ng-if="sizeOf(restoreVersions.versions) == 0" translate>There are no file versions to restore.</span>
+              </td>
```
I assume this exists for the time until fancytree is initialised, isn't it? What's the point of the conditional, instead of just always showing "loading data" until fancytree is ready?

tomasz1986 replied:
Good question, and the code was remnants of the previous iteration, however I've now added yet another similar `sizeOf` check in the controller, so that the tree code doesn't run at all when there are no versions. This way, the window loads faster, there is no 1970 year in the disabled date filter, and the two conditionals are actually required again (because without the tree code running, the existing table isn't replaced with the tree table).

I hope the explanation is clear… What do you think?

imsodin replied:
Yep makes sense. I don't think I'll remember/be able to figure it out if I ever stumble over this again though, but it wasn't any better before so 🤷

imsodin reviewed (commented):
Looks good, just one change I don't understand.
